csharp/src/Ice/Communicator-LocatorManager.cs

```diff
@@ -12,437 +13,366 @@ namespace ZeroC.Ice
+                {
+                    _table.RemoveAdapterEndpoints(reference.AdapterId);
+                    throw;
                 }
-                catch (System.Exception ex)
+                finally
                 {
-                    Exception(ex);
+                    lock (_mutex)
+                    {
+                        _adapterRequests.Remove(reference.AdapterId);
+                    }
                 }
             }
         }
 
-        internal LocatorInfo(ILocatorPrx locator, LocatorTable table, bool background)
-        {
-            Locator = locator;
-            _table = table;
-            _background = background;
-        }
-
-        internal void Destroy()
-        {
-            lock (_mutex)
-            {
-                _locatorRegistry = null;
-                _table.Clear();
-            }
-        }
+        internal ILocatorRegistryPrx? GetLocatorRegistry() => _locatorRegistry.Value;
 
-        public override bool Equals(object? obj)
+        private async Task<Reference?> GetObjectReferenceAsync(Reference reference)
         {
-            if (ReferenceEquals(this, obj))
+            if (reference.Communicator.TraceLevels.Location > 0)
             {
-                return true;
+                reference.Communicator.Logger.Trace(reference.Communicator.TraceLevels.LocationCat,
+                    $"searching for well-known object\nwell-known proxy = {reference}");
             }
 
-            return !(obj is LocatorInfo rhs) ? false : Locator.Equals(rhs.Locator);
-        }
-
-        public override int GetHashCode() => Locator.GetHashCode();
-
-        // No synchronization necessary, _locator is immutable.
-        internal ILocatorPrx Locator { get; }
-
-        internal ILocatorRegistryPrx? GetLocatorRegistry()
-        {
+            Task<Reference?>? task;
             lock (_mutex)
             {
-                if (_locatorRegistry != null)
+                if (!_objectRequests.TryGetValue(reference.Identity, out task))
                 {
-                    return _locatorRegistry;
+                    // If there's no locator request in progress for this object, we make one and cache it to prevent
+                    // making too many requests on the locator. It's removed once the locator response is received.
+                    task = PerformGetObjectProxyAsync(reference);
+                    if (!task.IsCompleted)
+                    {
+                        _objectRequests.Add(reference.Identity, task);
+                    }
                 }
             }
 
-            //
-            // Do not make locator calls from within sync.
-            //
-            ILocatorRegistryPrx? locatorRegistry = Locator.GetRegistry();
-            if (locatorRegistry == null)
+            try
             {
-                return null;
+                // Wait and return the locator response.
+                return await task.ConfigureAwait(false);
```

Comment:
Could we simply `return task` and move the exception tracing to `PerformGetAdapterEndpointsAsync`?

Reply:
Yes, we can do that. The exception handling is actually bogus here, we catch AdapterNotFoundException instead of ObjectNotFoundException... I'll fix it.